compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp

```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
+  }
+
+  if (!isa<linalg::MatmulOp, linalg::BatchMatmulOp>(op)) {
+    return success();
+  }
+  LLVM_DEBUG({
+    llvm::dbgs() << "verifying op: " << *op << "\n";
+    llvm::dbgs() << "chosen workgroup size: [";
+    llvm::interleaveComma(workgroupSize, llvm::dbgs());
+    llvm::dbgs() << "]\n";
+  });
+
+  // Get spirv.target_env attributes
+  const spirv::TargetEnvAttr targetEnvAttr = getSPIRVTargetEnvAttr(op);
+  const spirv::TargetEnv targetEnv(targetEnvAttr);
+  const auto limits = targetEnv.getResourceLimits();
+  LLVM_DEBUG(llvm::dbgs() << "target environment: " << targetEnvAttr << "\n");
+
+  const int subgroupSize = limits.getSubgroupSize();
+  const int maxSharedMemory = limits.getMaxComputeSharedMemorySize();
+  const int maxThreads = limits.getMaxComputeWorkgroupInvocations();
+  const auto maxWorkGroupSize = llvm::to_vector<3>(llvm::map_range(
+      limits.getMaxComputeWorkgroupSize().getAsValueRange<IntegerAttr>(),
+      [](const APInt &dim) { return dim.getSExtValue(); }));
+
+  // Verify each dimension of workgroupSize should be power of two
+  if (!llvm::isPowerOf2_64(workgroupSize[0]) ||
+      !llvm::isPowerOf2_64(workgroupSize[1]) ||
+      !llvm::isPowerOf2_64(workgroupSize[2])) {
+    return op->emitOpError(
+        "expected each workgroup size dimension to be power of two");
+  }
+
+  // Verify each dimension of workgroup size should not exceed the corresponding
+  // limit of maxWorkGroupSize
+  if (workgroupSize[0] > maxWorkGroupSize[0] ||
+      workgroupSize[1] > maxWorkGroupSize[1] ||
+      workgroupSize[2] > maxWorkGroupSize[2]) {
+    return op->emitOpError("expected workgroup size dimensions not exceeding ")
+           << "[" << maxWorkGroupSize[0] << ", " << maxWorkGroupSize[1] << ", "
+           << maxWorkGroupSize[2] << "]";
+  }
+
+  // Verify the total workgroup size should not exceed maxThreads
+  int64_t totalWorkgroupSize =
+      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+  if (totalWorkgroupSize > maxThreads) {
+    return op->emitOpError(
+               "expected total invocation count in workgroup to be <= ")
+           << maxThreads << ", got " << totalWorkgroupSize;
+  }
+
+  // Verify the total workgroup size should be multiple of subgroupSize
+  if (totalWorkgroupSize % subgroupSize != 0) {
+    return op->emitOpError("expected total workgroup size to be multiple of ")
+           << subgroupSize;
+  }
+
+  // Verify the total workgroup size should be equal or larger than 2 *
+  // subgroupSize
+  if (totalWorkgroupSize / subgroupSize < 2) {
+    return op->emitOpError("expected total workgroup size to be >= ")
+           << 2 * subgroupSize;
+  }
+
+  Type inputType = op->getOperand(0).getType();
+  ArrayRef<int64_t> lhsShape =
+      op->getOperand(0).getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> rhsShape =
+      op->getOperand(1).getType().cast<ShapedType>().getShape();
+
+  SmallVector<int64_t> firstLevelTileSizes =
```

Comment:
s/firstLevel/workgrup/?

Reply:
Do you mean to change the variable name from firstLevelTileSizes to something like workgroupTileSizes?

Reply:
Yup!